core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -1589,6 +1589,67 @@ private void checkPredicates(RelOptCluster cluster, RelOptTable empTable,
         sortsAs("[IS NULL($0)]"));
   }
 
+  @Test public void testPullUpPredicatesFromUnion() {
+    final RelMetadataQuery mq = RelMetadataQuery.instance();
+    RelNode rel = null;
+    rel = convertSql(""
+        + "select empno from emp where empno=1\n"
+        + "union all\n"
+        + "select empno from emp where empno=1");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+
+    rel = convertSql(""
+        + "select empno, deptno from emp where empno=1 or deptno=2\n"
+        + "union all\n"
+        + "select empno, deptno from emp where empno=3 or deptno=4");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[OR(=($0, 1), =($1, 2), =($0, 3), =($1, 4))]"));
+
+    rel = convertSql(""
+        + "select empno, comm, deptno from emp where empno=1 and comm=2 and deptno=3\n"
+        + "union all\n"
+        + "select empno, comm, deptno from emp where empno=1 and comm=4");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1), OR(AND(=($1, 2), =($2, 3)), =($1, 4))]"));
+
+  }
+
+  @Test public void testPullUpPredicatesFromIntersect() {
+    final RelMetadataQuery mq = RelMetadataQuery.instance();
+    RelNode rel = null;
+    rel = convertSql(""
+        + "select empno from emp where empno=1\n"
+        + "intersect all\n"
+        + "select empno from emp where empno=1");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+
+    rel = convertSql(""
+        + "select empno, deptno from emp where empno=1 and deptno=2\n"
+        + "intersect all\n"
+        + "select empno, deptno from emp where empno=1 and deptno=3");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+
+    rel = convertSql(""
+        + "select empno, deptno, comm from emp where 1=empno and deptno=2\n"
+        + "intersect all\n"
+        + "select empno, deptno, comm from emp where empno=1 and comm=3");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+  }
+
+  @Test public void testPullUpPredicatesFromMinus() {
+    final RelMetadataQuery mq = RelMetadataQuery.instance();
+    RelNode rel = convertSql(""
+        + "select empno, deptno, comm from emp where empno=1 and deptno=2\n"
+        + "except all\n"
+        + "select empno, deptno, comm from emp where comm=3");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1), =($1, 2)]"));
```

Comment:
If it is except, it should also have `NOT(comm=3)`. Can you add this case?

Reply:
True in this case (because the two branches have the same FROM clause) but I think it may be difficult to deduce in general.

For the case that the branches have the same FROM clause, we'd be better served to have a rewrite to 'select ... from emp where empno = 1 and deptno = 2 and not (comm = 3)'. I wonder how often that rewrite would apply in practice.

Reply:
Yes, given the same FROM clause, the perfect answer for this case would be `empno=1 and deptno=2 and not(comm=3)`.
But can I keep the logic simple and just return `empno=1 and deptno=2` ?

Reply:
Yes. No problem. I don't think it applies often.

Reply:
> True in this case (because the two branches have the same FROM clause) but I think it may be difficult to deduce in general.

@julianhyde  Even if they are different FROM clauses, it still can have common expression of `NOT($2=3)`, the source doesn't really matter.
But I agree, the optimization doesn't apply often.

Reply:
I'm thinking of
```
select * from employees
except
select * from managers where deptno = 10
```
where `employees` and `managers` have the same fields (but not the same rows). I think you'll agree that the output does not obey the constraint `NOT (deptno = 10)`, because the inputs have different FROM clauses.

Reply:
Yes, that's right ~

Reply:
Hmm, after a second thought, I think the output doesn't obey the  constraint `NOT (deptno = 10)`, only when the second child `(select * from managers where deptno = 10)` output 0 tuples. Otherwise, it still obeys. But we don't know it is empty or not, so we can't really derive condition from second child.